### ISSUE TITLE
Improved queries

### ISSUE THREAD

**Issue by peter-conalgo** [josiahcarlson/rom#66]
Feature request/suggestion... 

- We should be able `query.count()` any model that has a unique field. It's just a question of finding the right column (...`uidx`), and doing an `HLEN` on the hash for that column.

- Similarly we should be able to iterate over all members in any model, using the newer (v 2.8) [HSCAN](http://redis.io/commands/hscan) redis command. The order of results would be random and some items maybe missed if the table changes during operation, but it would still be useful.

- Even for models with no index and no unique fields, we can still use [SCAN](http://redis.io/commands/scan) to count and index records. A runtime warning about poor performance/bad design might be in order, but it can still work.

Bottom line: I'd really like to see `Model.count()` and `for i in Model` just work. 

(BTW: Big fan of ROM.)

**Comment by josiahcarlson** [comment]
Before I respond to your specific feature requests, let me explain a few things. Hashes in Redis are very limited in what they can do, and don't play nicely with sets or zsets. Things like `query.filter()`, `.like()`, `.startswith()`, `.endswith()`, `.order_by()`, and `.limit()` uses sets or zsets in a variety of ways to turn the result into a zset - which is trivially paginated over with `ZRANGE`. Hashes don't have arbitrary "give me 30 items starting from offset n/2" (though like you say, `ZSCAN` does let you incrementally scan a hash, and it is conditionally used in at least one place in rom already), so a typical `.limit()` clause won't work. The only case that makes sense is if someone were to do something like `Model.query.order_by('unique_index').iter_result()`

* Re: Using unique indexes for `query.count()`

`None` values are not stored in the unique index, and do not count as unique values. So any entity that has `None` as a value in a unique indexed column wouldn't be counted by a `query.count()`. Requiring `required=True` might be okay, but there is something else that can be used.

A little bit deeper, there is a separate hash that actually has a mapping from all entity ids to the json encoded version of raw index data (used for un-indexing to ensure that all known index data is cleared). That said, some fairly old versions of rom didn't clean this index data properly, and other users like to delete the underlying hash key that stores the rom entity itself, so in some cases *that* result could be incorrect too.

To offer `Model.query.count()` functionality, it may be sufficient to document that users upgrading from very old versions of rom or those deleting keys directly (instead of using `entity.delete()`) will need to run `all(rom.util.clean_old_index(model))` after upgrading/deleting in order to get a correct answer. I can add this.

* Re: HSCAN

Using HSCAN only really makes sense if someone uses `Model.query.iter_result()` to iterate over the results, as allowing `Model.query.all()` is a bad idea.

* Re: SCAN

SCAN is unnecessary, see the hash of id -> index data mentioned earlier.

I'm unsure what you mean by "and index records". If you mean refreshing indexes, rom already has a utility function for re-indexing data if you've changed your index fields, it's called `util.refresh_indices(model)`, and can be used like `all(util.refresh_indices(model))` .This doesn't use HSCAN simply because there is a very good chance that for any nontrivial rom use, the values in *any* index can change during iteration, so we can't rely on them. Right now it actually uses `Model.get([<a range of ids that may or may not exist>])`, which could be improved.

* Re: `Model.count()` and `for i in Model`

What if it was spelled `Model.query.count()` and `for i in Model.query.iter_results()` or even `for i in Model.query`? I ask because right now all of the "get me information about this group of things" interface is hanging off the `Model.query` attribute, and I can foresee someone already utilizing the `.count` attribute on a model for existing data. No such limitation or concern with modifying `Model.query`.

**Comment by peter-conalgo** [comment]
Sorry, yes, I meant: `Model.query.count()` and `Model.query.iter_results()`. 

At the same time, would be nice if `Model.query` itself would operate as an iterable as you mentioned, so idiom: `for i in Model.query` works.

**Comment by josiahcarlson** [comment]
I've got an implementation for both of these, plus a bigfix, but I'm going to wait until tomorrow to release it. I ended up not using HSCAN for iterating over a bare `Model.query` simply because modifications of entities during iteration could result in the same entity being returned twice - which would be obviously wrong.

**Comment by josiahcarlson** [comment]
These queries are now available in Github as tag 0.32.0 and from PyPI as version 0.32.0 . Thank you for the feature request :)
